### PR TITLE
Add sampling to synthesizers

### PR DESCRIPTION
**Pull request by pvk-developer**
Resolves #1050

### CODE REVIEW

File: sdv/single_table/base.py [hunk 1]
```diff
@@ -75,3 +95,638 @@ def fit(self, data):
...
+                * int:
+                    Number of rows that are considered valid.
+        """
+        if self._data_processor._dtypes is not None:
+            if conditions is None:
+                sampled = self._sample(num_rows)
+            else:
+                try:
+                    sampled = self._sample(num_rows, transformed_conditions)
+                except TypeError:
```
Comment (amontanez24) [inline sdv/single_table/base.py:173]: why did we change this? This is for the case when certain models don't support conditional sampling so they raise a `NotImplementedError`. For example, `TVAE` and `CTGAN`

File: tests/unit/single_table/test_base.py [hunk 2]
```diff
@@ -104,3 +108,584 @@ def test_fit(self):
...
+        data = pd.DataFrame({
+            'name': ['John', 'Doe', 'John Doe'],
+            'salary': [90.0, 100.0, 80.0]
+        })
+        instance = Mock()
+        instance._data_processor._dtypes = pd.Series()
+        instance._filter_conditions.return_value = data[data.name == 'John Doe']
+        conditions = {'salary': 80.}
+        transformed_conditions = {'salary.value': 80.0}
+        instance._sample.side_effect = [TypeError, None]
```
Comment (amontanez24) [inline tests/unit/single_table/test_base.py:247]: this should be changed to a `NotImplementedError`, or the code in `CTGAN` and `TVAE` needs to change to raise a `TypeError`

File: tests/unit/single_table/test_base.py [hunk 3]
```diff
@@ -104,3 +108,584 @@ def test_fit(self):
...
+        instance._sample.assert_called_once_with(3, {'salary.value': 80.0})
+        instance._data_processor.reverse_transform.assert_called_once_with(
+            instance._sample.return_value
+        )
+        instance._data_processor.filter_valid.assert_called_once_with(
+            instance._data_processor.reverse_transform.return_value
+        )
+
+    def test__sample_rows_with_previous_rows(self):
+        """Test that when calling sample rows with previous rows those are being concatenated."""
```
Comment (amontanez24) [inline tests/unit/single_table/test_base.py:204]: minor: can we say `Test that previous rows are being concatenated when provided to `_sample`.

File: sdv/single_table/base.py [hunk 4]
```diff
@@ -75,3 +95,638 @@ def fit(self, data):
...
+                sampled = self._filter_conditions(sampled, conditions, float_rtol)
+
+            num_valid = len(sampled)
+
+            return sampled, num_valid
+
+        else:
+            sampled = pd.DataFrame(index=range(num_rows))
+            sampled = self._data_processor.reverse_transform(sampled)
+            return sampled, num_rows
```
Comment (amontanez24) [inline sdv/single_table/base.py:193]: Do we know what this situation is for? Is this for anonymization?

File: tests/unit/single_table/test_base.py [hunk 5]
```diff
@@ -104,3 +108,584 @@ def test_fit(self):
...
+
+        # Run
+        sampled, num_rows = BaseSynthesizer._sample_rows(instance, 10)
+
+        # Assert
+        assert num_rows == 10
+        pd.testing.assert_frame_equal(sampled, pd.DataFrame(index=range(10)))
+
+    def test__sample_batch_without_saving_to_file(self):
+        # Setup
```
Comment (amontanez24) [inline tests/unit/single_table/test_base.py:277]: missing docstring

File: tests/unit/single_table/test_base.py [hunk 6]
```diff
@@ -104,3 +108,584 @@ def test_fit(self):
...
+            output_file_path=None,
+        )
+
+        # Assert
+        pd.testing.assert_frame_equal(result, sampled_data)
+        rows, conditions, trans_cond, float_rtol, sampled = instance._sample_rows.call_args[0]
+        assert instance._sample_rows.call_count == 2
+
+    def test__sample_batch_storing_output_file(self, tmpdir):
+        """Test that an output file is properly stored while sampling and progress bar updated."""
```
Comment (amontanez24) [inline tests/unit/single_table/test_base.py:369]: minor: can we rephrase to `Test that an output file is properly stored while sampling and that the progress bar is updated`

File: tests/unit/single_table/test_base.py [hunk 7]
```diff
@@ -104,3 +108,584 @@ def test_fit(self):
...
+            max_tries_per_batch=None,
+            conditions=condition,
+            transformed_conditions=transformed_condition,
+            float_rtol=0.01,
+            progress_bar=None,
+            output_file_path=None
+        )
+
+    def test__conditionally_sample_rows_raises_value_error(self):
+        """Test when sampled rows is lower or 0."""
```
Comment (amontanez24) [inline tests/unit/single_table/test_base.py:537]: for all of these I'd reword to `when no rows are sampled`. I think we can also add more description to explain that this is when no rows are sampled and `graceful_reject_sampling` is `False`

File: tests/unit/single_table/test_base.py [hunk 8]
```diff
@@ -104,3 +108,584 @@ def test_fit(self):
...
+            batch_size=3,
+            max_tries_per_batch=None,
+            conditions=condition,
+            transformed_conditions=transformed_condition,
+            float_rtol=0.01,
+            progress_bar=None,
+            output_file_path=None
+        )
+
+    def test__conditionally_sample_rows_raises_value_error(self):
```
Comment (amontanez24) [inline tests/unit/single_table/test_base.py:536]: I think the name here could describe why the error is raise, eg. `test__conditionally_sample_rows_no_rows_sampled_error`

File: tests/unit/single_table/test_base.py [hunk 9]
```diff
@@ -104,3 +108,584 @@ def test_fit(self):
...
+        with pytest.raises(ValueError, match=expected_message):
+            BaseSynthesizer._conditionally_sample_rows(
+                instance,
+                transformed_data,
+                condition,
+                transformed_condition,
+                graceful_reject_sampling=False,
+            )
+
+    def test__conditionally_sample_rows_without_grafecul_reject_sampling(self):
```
Comment (amontanez24) [inline tests/unit/single_table/test_base.py:592]: shouldn't this say `with_graceful_reject_sampling` since it is `True` in this test. also a typo: grafecul

File: tests/unit/single_table/test_base.py [hunk 10]
```diff
@@ -104,3 +108,584 @@ def test_fit(self):
...
+            condition,
+            transformed_condition,
+            graceful_reject_sampling=True,
+        )
+
+        # Assert
+        assert result == []
+
+    def test__randomize_samples_false(self):
+        """Test when the ``randomize_samples`` parameter is ``False``."""
```
Comment (amontanez24) [inline tests/unit/single_table/test_base.py:618]: for both of these tests I think we can add one sentence to explain the desired behavior

**Review by amontanez24 — CHANGES_REQUESTED** [review]
This is looking good so far. I left a few comments about rewording and adding more explanation to the docs. I also think there is a mistake with one of the errors we're catching but besides that it looks good so far

File: sdv/single_table/base.py (hunk 1 above)
Comment (pvk-developer) [reply sdv/single_table/base.py:173]: Good catch, I thought it was a `TypeError` because we basically didn't accept that extra argument on those models (as we call the same method below).

File: sdv/single_table/base.py (hunk 4 above)
Comment (pvk-developer) [reply sdv/single_table/base.py:193]: This is one of the open questions I had. I don't see when `data_processor` will not have `Dtypes` 🤔  (When not fitted?)

Comment (amontanez24) [reply sdv/single_table/base.py:193]: Try running through the integration tests for `sdv/tabular/base.py` and see if any reach this point

Comment (pvk-developer) [reply sdv/single_table/base.py:193]: This case is when there are no other columns but `primary_keys` or `ids` . We will have to make some changes on the `data_processor` to get this case allowed.

File: sdv/single_table/copulas.py [hunk 11]
```diff
@@ -108,3 +108,29 @@ def _fit(self, processed_data):
...
             self._model.fit(processed_data)
+
+    def _set_random_state(self, random_state):
+        """Set the random state of the model's random number generator.
+
+        Args:
+            random_state (int, np.random.RandomState, or None):
+                Seed or RandomState to use.
+        """
+        self._model.set_random_state(random_state)
```
[the remark this answers is not in the document]
Comment (pvk-developer) [reply sdv/single_table/copulas.py:119]: I don't see why not.

File: tests/unit/single_table/test_base.py [hunk 12]
```diff
@@ -104,3 +109,982 @@ def test_fit(self):
...
+            transformed_conditions=transformed_conditions
+        )
+
+        # Assert
+        assert num_valid == 1
+        pd.testing.assert_frame_equal(sampled, data[data.name == 'John Doe'])
+        assert instance._sample.call_args_list == [call(3, {'salary.value': 80.0}), call(3)]
+
+    def test__sample_rows_dtypes_is_none(self):
+        """Test when ``_data_processor._dtypes`` is ``None``."""
```
Comment (amontanez24) [inline tests/unit/single_table/test_base.py:267]: maybe add the context here that this is for when there are only primary keys

File: tests/unit/single_table/test_base.py [hunk 13]
```diff
@@ -104,3 +109,982 @@ def test_fit(self):
...
+            BaseSynthesizer._conditionally_sample_rows(
+                instance,
+                transformed_data,
+                condition,
+                transformed_condition,
+                graceful_reject_sampling=False,
+            )
+
+    def test__conditionally_sample_rows_raises_value_error_model_is_gm(self):
+        """Test when sampled rows is lower or 0 and model is ``GaussianMultivariate``."""
```
Comment (amontanez24) [inline tests/unit/single_table/test_base.py:573]: can we improve this docstring to say testing when there are no sampled rows and the model is GaussianMultivariate. Also, wheat is supposed to happen in this test?

File: tests/unit/single_table/test_base.py [hunk 14]
```diff
@@ -104,3 +109,982 @@ def test_fit(self):
...
+            BaseSynthesizer._conditionally_sample_rows(
+                instance,
+                transformed_data,
+                condition,
+                transformed_condition,
+                graceful_reject_sampling=False,
+            )
+
+    def test__conditionally_sample_rows_with_graceful_reject_sampling(self):
+        """Test when sampled rows is lower or 0 but ``graceful_reject_sampling`` is ``True``."""
```
Comment (amontanez24) [inline tests/unit/single_table/test_base.py:601]: minor: change lower or 0 to no rows are sampled

File: tests/unit/single_table/test_base.py [hunk 15]
```diff
@@ -561,3 +563,990 @@ def test_update_transformers(self):
         assert len(field_transformers) == 2
         assert isinstance(field_transformers['col1'], GaussianNormalizer)
         assert isinstance(field_transformers['col2'], GaussianNormalizer)
+
+    @patch('sdv.single_table.base.DataProcessor')
+    def test__set_random_state(self, mock_data_processor):
+        """Test that error is raised when this is not implemented."""
```
Comment (fealho) [inline tests/unit/single_table/test_base.py:569]: Wrong description

File: sdv/single_table/base.py [hunk 16]
```diff
@@ -276,3 +294,648 @@ def update_transformers(self, column_name_to_transformer):
...
+
+        Returns:
+            pandas.DataFrame:
+                Rows from the sampled data that match the conditions.
+        """
+        for column, value in conditions.items():
+            column_values = sampled[column]
+            if column_values.dtype.kind == 'f':
+                distance = value * float_rtol
+                sampled = sampled[np.abs(column_values - value) <= distance]
```
Comment (fealho) [inline sdv/single_table/base.py:330]: Not sure if this works with nans (if it's possible for nans to reach this stage).

File: tests/unit/single_table/test_base.py [hunk 17]
```diff
@@ -561,3 +563,990 @@ def test_update_transformers(self):
...
+        instance = BaseSynthesizer(metadata)
+        instance._model = Mock()
+
+        # Run
+        instance._set_random_state(rng_seed)
+
+        # Assert
+        instance._model.set_random_state.assert_called_once_with(rng_seed)
+
+    def test__filter_conditions(self):
```
Comment (fealho) [inline tests/unit/single_table/test_base.py:582]: Maybe add a few more `'Software Engineer'` columns and `80.` columns, to make sure it's correctly conditioning on both of them (rn if it only conditions on one of them, the result would be the same)

File: tests/unit/single_table/test_base.py [hunk 18]
```diff
@@ -561,3 +563,990 @@ def test_update_transformers(self):
...
+
+        # Assert
+        assert num_valid == 1
+        pd.testing.assert_frame_equal(sampled, data[data.name == 'John Doe'])
+        assert instance._sample.call_args_list == [call(3, {'salary.value': 80.0}), call(3)]
+
+    def test__sample_rows_dtypes_is_none(self):
+        """Test when ``_data_processor._dtypes`` is ``None``.
+
+        This test is when ``_data_processor._dtypes`` is ``None``, wich by the legacy
```
Comment (fealho) [inline tests/unit/single_table/test_base.py:723]: which

File: tests/unit/single_table/test_base.py [hunk 19]
```diff
@@ -561,3 +563,990 @@ def test_update_transformers(self):
...
+        code was when only ``ids`` or ``primary_keys`` where sampled.
```
Comment (fealho) [inline tests/unit/single_table/test_base.py:724]: were

File: sdv/single_table/base.py [hunk 20]
```diff
@@ -276,3 +294,648 @@ def update_transformers(self, column_name_to_transformer):
...
+        """
+        num_rows_to_sample = batch_size
+
+        counter = 0
+        num_valid = 0
+        prev_num_valid = None
+        remaining = batch_size
+        sampled = pd.DataFrame()
+
+        while num_valid < batch_size:
```
Comment (fealho) [inline sdv/single_table/base.py:453]: Rewrite as `while num_valid < batch_size and counter < max_tries`

File: sdv/single_table/base.py [hunk 21]
```diff
@@ -276,3 +294,648 @@ def update_transformers(self, column_name_to_transformer):
...
+            if counter >= max_tries:
```
Comment (fealho) [inline sdv/single_table/base.py:454]: Shouldn't we raise a warning if this happens, like we do in CTGAN?

File: tests/unit/single_table/test_base.py [hunk 22]
```diff
@@ -561,3 +563,990 @@ def test_update_transformers(self):
...
+        instance._sample_rows.side_effect = [
+            (sampled_data, 4),
+            (sampled_data, 5),
+            (sampled_data, 10),
+        ]
+        mock_progress_bar = Mock()
+
+        path = tmpdir / 'file.csv'
+        # Create an empty csv file
+        open(path, mode='a').close()
```
Comment (fealho) [inline tests/unit/single_table/test_base.py:851]: Never seen anyone do this before 😆

File: tests/unit/single_table/test_base.py [hunk 23]
```diff
@@ -561,3 +563,990 @@ def test_update_transformers(self):
...
+        # Assert
+        expected_result = [
+            pd.DataFrame({'name': ['John Doe']}),
+            pd.DataFrame({'salary': [80.]}),
+        ]
+
+        for res, exp in zip(result, expected_result):
+            pd.testing.assert_frame_equal(res, exp)
+
+    def test__sample_in_batches(self):
```
Comment (fealho) [inline tests/unit/single_table/test_base.py:914]: Need another test where result is empty (to get 100% coverage).

**Review by fealho — COMMENTED** [review]
First batch of reviews

File: sdv/single_table/base.py [hunk 24]
```diff
@@ -276,3 +294,648 @@ def update_transformers(self, column_name_to_transformer):
...
+            transformed_conditions=transformed_condition,
+            float_rtol=float_rtol,
+            progress_bar=progress_bar,
+            output_file_path=output_file_path
+        )
+
+        if len(sampled_rows) > 0:
+            sampled_rows[COND_IDX] = dataframe[COND_IDX].to_numpy()[:len(sampled_rows)]
+
+        else:
```
Comment (fealho) [inline sdv/single_table/base.py:558]: do `elif not graceful_reject_sampling:`

File: tests/unit/single_table/test_base.py [hunk 25]
```diff
@@ -561,3 +563,990 @@ def test_update_transformers(self):
...
+    def test__sample_with_progress_bar_with_conditions(self):
+        """Test that a ``TypeError`` is raised when there are conditions."""
+        # Setup
+        instance = Mock()
+        conditions = [Mock(), Mock()]
+
+        # Run and Assert
+        expected_message = re.escape(
+            'This method does not support the conditions parameter. '
+            'Please create `sdv.sampling.Condition` objects and pass them '
```
Comment (fealho) [inline tests/unit/single_table/test_base.py:1139]: We use ' instead of ` to highlight objects in strings

File: tests/unit/single_table/test_base.py [hunk 26]
```diff
@@ -561,3 +563,990 @@ def test_update_transformers(self):
...
+            'Sampling rows'
+        )
+        instance._sample_in_batches.assert_called_once_with(
+            num_rows=10,
+            batch_size=10,
+            max_tries_per_batch=100,
+            progress_bar=progress_bar.__enter__.return_value,
+            output_file_path=mock_validate_file_path.return_value,
+        )
+        mock_handle_sampling_error(False, 'temp_file', KeyboardInterrupt())
```
Comment (fealho) [inline tests/unit/single_table/test_base.py:1233]: Not sure what this line does?

File: tests/unit/single_table/test_base.py [hunk 27]
```diff
@@ -561,3 +563,990 @@ def test_update_transformers(self):
...
+    def test__validate_conditions_raises_error(self):
+        """Test that conditions are not in the ``data_processor`` fields."""
+        # Setup
+        instance = Mock()
+        instance._data_processor.get_fields.return_value = ['name', 'surname']
+        conditions = pd.DataFrame({'name.value': ['Johanna'], 'surname.value': ['Doe']})
+
+        # Run and Assert
+        error_msg = re.escape(
+            'Unexpected column name `name.value`. Use a column name that was present in the '
```
Comment (fealho) [inline tests/unit/single_table/test_base.py:1326]: Same, use ' instead of `

File: tests/unit/single_table/test_base.py [hunk 28]
```diff
@@ -561,3 +563,990 @@ def test_update_transformers(self):
...
+        # Setup
+        conditions = pd.DataFrame({
+            'name': ['Johanna', 'Doe'],
+            'salary': [100., 90.]
+        })
+        instance = Mock()
+        instance._data_processor.transform.side_effect = [ConstraintsNotMetError]
+
+        # Run and Assert
+        error_msg = 'Provided conditions are not valid for the given constraints'
```
Comment (fealho) [inline tests/unit/single_table/test_base.py:1343]: End sentence with a dot.

File: sdv/single_table/base.py [hunk 29]
```diff
@@ -276,3 +294,648 @@ def update_transformers(self, column_name_to_transformer):
...
+            if output_file_path == TMP_FILE_NAME and os.path.exists(output_file_path):
+                os.remove(output_file_path)
+
+        return sampled
+
+    def sample_remaining_columns(self, known_columns, max_tries_per_batch=100, batch_size=None,
+                                 randomize_samples=True, output_file_path=None):
+        """Sample rows from this table.
```
Comment (fealho) [inline sdv/single_table/base.py:905]: I think this needs a little more description.

**Review by fealho — COMMENTED** [review]
This looks great! I left a few more minor nitpicks, but things look pretty 👍 

Only thing I think is missing is some integration tests, it would set my heart at rest knowing things work end-to-end.

File: tests/unit/single_table/test_base.py (hunk 26 above)
Comment (pvk-developer) [reply tests/unit/single_table/test_base.py:1233]: missing `assert_called_with` :D thanks!

File: sdv/single_table/base.py (hunk 21 above)
Comment (pvk-developer) [reply sdv/single_table/base.py:454]: @amontanez24 may have the proper answer for this.

File: sdv/single_table/base.py (hunk 24 above)
Comment (pvk-developer) [reply sdv/single_table/base.py:558]: nice

File: tests/unit/single_table/test_base.py (hunk 23 above)
Comment (pvk-developer) [reply tests/unit/single_table/test_base.py:926]: 👍🏻  done

File: tests/unit/data_processing/test_data_processor.py [hunk 30]
```diff
@@ -293,6 +293,53 @@ def test_set_model_kwargs(self):
         # Assert
         assert dp._model_kwargs == {'model': {'arg1': 10, 'arg2': True}}
 
+    def test_get_sdtypes(self):
+        """Test that this returns a mapping of column names and it's sdtypes.
```
Comment (fealho) [inline tests/unit/data_processing/test_data_processor.py:297]: its, not it's

File: sdv/single_table/base.py (hunk 16 above)
Comment (fealho) [reply sdv/single_table/base.py:332]: @pvk-developer Did you get a chance to check this?

File: sdv/single_table/base.py (hunk 21 above)
Comment (amontanez24) [reply sdv/single_table/base.py]: I don't think `CTGAN` has any different warning messages for sampling. Which one are you referring to?

File: sdv/single_table/ctgan.py [hunk 31]
```diff
@@ -183,3 +203,23 @@ def _fit(self, processed_data):
...
+
+    def _sample(self, num_rows, conditions=None):
+        """Sample the indicated number of rows from the model.
+
+        Args:
+            num_rows (int):
+                Amount of rows to sample.
+            conditions (dict):
+                If specified, this dictionary maps column names to the column
+                value. Then, this method generates `num_rows` samples, all of
```
Comment (amontanez24) [inline sdv/single_table/ctgan.py:215]: should this have double `?

File: sdv/single_table/ctgan.py [hunk 32]
```diff
@@ -111,6 +111,26 @@ def _fit(self, processed_data):
...
 
+    def _sample(self, num_rows, conditions=None):
+        """Sample the indicated number of rows from the model.
+
+        Args:
+            num_rows (int):
+                Amount of rows to sample.
+            conditions (dict):
+                If specified, this dictionary maps column names to the column
+                value. Then, this method generates `num_rows` samples, all of
```
Comment (amontanez24) [inline sdv/single_table/ctgan.py:122]: should this have double ` around num_rows?

File: sdv/single_table/copulas.py [hunk 33]
```diff
@@ -124,3 +124,20 @@ def _warn_for_update_transformers(self, column_name_to_transformer):
...
+
+    def _sample(self, num_rows, conditions=None):
+        """Sample the indicated number of rows from the model.
+
+        Args:
+            num_rows (int):
+                Amount of rows to sample.
+            conditions (dict):
+                If specified, this dictionary maps column names to the column
+                value. Then, this method generates `num_rows` samples, all of
```
Comment (amontanez24) [inline sdv/single_table/copulas.py:136]: same

File: sdv/single_table/copulagan.py [hunk 34]
```diff
@@ -184,3 +184,21 @@ def _fit(self, processed_data):
...
+
+    def _sample(self, num_rows, conditions=None):
+        """Sample the indicated number of rows from the model.
+
+        Args:
+            num_rows (int):
+                Amount of rows to sample.
+            conditions (dict):
+                If specified, this dictionary maps column names to the column
+                value. Then, this method generates `num_rows` samples, all of
```
Comment (amontanez24) [inline sdv/single_table/copulagan.py:196]: same

File: sdv/single_table/base.py (hunk 21 above)
Comment (amontanez24) [reply sdv/single_table/base.py]: The  warning for not being able to sample enough rows surfaces later down the pipeline

File: sdv/single_table/base.py (hunk 16 above)
Comment (pvk-developer) [reply sdv/single_table/base.py:332]: This returns empty  `pd.DataFrame` if by any chance this occurs or reaches this point, which I couldn't make reach.

Comment (amontanez24) [reply sdv/single_table/base.py:332]: This is how the current implementation is as well. I think it's fine for now. We should keep an eye on any bugs for this scenario

**Review by amontanez24 — APPROVED** [review]
I think this is ready to go 🎉 Good work! :shipit:
